src/Tools/BuildValidator/LocalReferenceResolver.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.Extensions.Logging;
+
+namespace BuildValidator
+{
+    /// <summary>
+    /// Resolves references for a package by looking in local nuget and artifact
+    /// directories for Roslyn
+    /// </summary>
+    internal class LocalReferenceResolver : IMetadataReferenceResolver
+    {
+        private readonly Dictionary<Guid, string> _cache = new Dictionary<Guid, string>();
+        private readonly HashSet<DirectoryInfo> _indexDirectories = new HashSet<DirectoryInfo>();
+        private readonly ILogger _logger;
+
+        public LocalReferenceResolver(ILoggerFactory loggerFactory)
+        {
+            _logger = loggerFactory.CreateLogger<LocalReferenceResolver>();
+            _indexDirectories.Add(GetArtifactsDirectory());
+            _indexDirectories.Add(GetNugetCacheDirectory());
+        }
+
+        public static DirectoryInfo GetNugetCacheDirectory()
```

Comment:
I worry a bit about using the NuGet directory here. It doesn't seem strictly necessary here, it would seem that the contents of `bin` should be enough for this tool to operate. 

Using the NuGet directory means it's possible that the validator will succeed on one machine but fail on another. That is because the NuGet direcotry contains the binaries from the totality of all restorres on the machine, not just the Roslyn one. This should be relatively safe in our CI but would like to avoid this if it's possible to do so.

Reply:
There's a lot of binaries we don't output to our bin. Here's the first 20 from my local testing. I'm not sure how to resolve these correctly otherwise? I think the idea that it's safe for our CI and scoped there should be sufficient, and hopefully if there is an issue it's easily diagnosable on a machine. 

```
    [0]: "Microsoft.Bcl.AsyncInterfaces.dll"
    [1]: "Microsoft.CodeAnalysis.dll"
    [2]: "Microsoft.CodeAnalysis.VisualBasic.Workspaces.dll"
    [3]: "Microsoft.CodeAnalysis.Workspaces.dll"
    [4]: "Microsoft.CSharp.dll"
    [5]: "Microsoft.Win32.Registry.dll"
    [6]: "mscorlib.dll"
    [7]: "netstandard.dll"
    [8]: "PresentationCore.dll"
    [9]: "System.Buffers.dll"
    [10]: "System.ComponentModel.Composition.dll"
    [11]: "System.Configuration.dll"
    [12]: "System.Core.dll"
    [13]: "System.Data.DataSetExtensions.dll"
    [14]: "System.Data.dll"
    [15]: "System.dll"
    [16]: "System.Drawing.dll"
    [17]: "System.IO.Compression.FileSystem.dll"
    [18]: "System.IO.dll"
    [19]: "System.IO.Pipelines.dll"
```